### PR TITLE
Allow tide to actually trigger tests and merge PRs.

### PR DESCRIPTION
**Pull request by spxtr**
/area prow

### CODE REVIEW

**Comment by xiangpengzhao** [comment]
@stevekuznetsov your favorite `WIP` :)

ref: https://github.com/kubernetes/test-infra/pull/4572

**Comment by stevekuznetsov** [comment]
hahaha omg

**Comment by spxtr** [comment]
@rmmh @kargakis this should be ready to go.

File: prow/tide/tide_test.go [hunk 1]
```diff
@@ -567,3 +574,210 @@ func TestPickBatch(t *testing.T) {
...
+			t.Fatalf("Adding initial commit: %v", err)
+		}
+
+		sp := subpool{
+			org:    "o",
+			repo:   "r",
+			branch: "master",
+			sha:    "master",
+		}
+		for i := 0; i < 10; i++ {
```
Comment (kargakis) [inline prow/tide/tide_test.go]: leftover?

File: prow/tide/tide.go [hunk 2]
```diff
@@ -276,57 +283,130 @@ func (c *Controller) pickBatch(sp subpool) ([]pullRequest, error) {
...
+			if _, ok := err.(github.ModifiedHeadError); ok {
+				// This is a possible source of incorrect behavior. If someone
+				// modifies their PR as we try to merge it in a batch then we
+				// end up in an untested state. This is unlikely to cause any
+				// real problems.
+				c.Logger.WithError(err).Info("Merge failed: PR was modified.")
+			} else if _, ok = err.(github.UnmergablePRError); ok {
+				c.Logger.WithError(err).Warning("Merge failed: PR is unmergable. How did it pass tests?!")
+			} else {
+				return err
```
Comment (kargakis) [inline prow/tide/tide.go]: What kind of error this can be?

Comment (spxtr) [reply prow/tide/tide.go]: If GitHub is down then we can't really recover properly. Better to just cut out here and retry on the next sync loop. I covered the two cases from which we can recover above.

File: prow/tide/tide_test.go (hunk 1 above)
Comment (spxtr) [reply prow/tide/tide_test.go]: Yes, removed.

**Comment by kargakis** [comment]
Needs a rebase

/lgtm

File: prow/tide/tide.go (hunk 2 above)
Comment (kargakis) [reply prow/tide/tide.go]: ack

**Comment by xiangpengzhao** [comment]
Don't we label or comment `needs-rebase` when a PR needs rebase in test-infra?

**Comment by spxtr** [comment]
> Don't we label or comment `needs-rebase` when a PR needs rebase in test-infra?

No, that's a munger currently.

**Pull request merged**